cpp/src/arrow/io/readahead.h

```diff
@@ -0,0 +1,94 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_IO_READAHEAD_H
+#define ARROW_IO_READAHEAD_H
+
+#include <memory>
+#include <string>
+
+#include "arrow/io/interfaces.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class MemoryPool;
+class ResizableBuffer;
+class Status;
+
+namespace io {
+namespace internal {
+
+struct ARROW_EXPORT ReadaheadBuffer {
+  std::shared_ptr<ResizableBuffer> buffer;
+  int64_t left_padding;
+  int64_t right_padding;
+};
+
+class ARROW_EXPORT ReadaheadSpooler {
```

Comment:
I think it would be useful to have a version of this, if not this class itself, that implements the `InputStream` protocol. Then we can pass in our async readahead reader into a function or class that is none the wiser. If we don't do it now, let's make sure we have a JIRA about doing so. Not yet understanding the motivation for the padding, if not for that it would seem possible to implement this as a subclass of InputStream

Reply:
The padding is meant to allow fixing line boundaries in CSV without having to copy the whole block.

Reply:
In any case, I don't think this is compatible with `InputStream`. The `InputStream` API requires that you can choose read size when `Read()` is called, but for non-wasteful readahead you want read sizes to be known in advance.